File: python/2022/05.py
from collections import defaultdict, namedtuple

Movement = namedtuple("Movement", ["count", "from_", "to"])


def process_data(lines: list[str]):
    iterator = iter(lines)

    index = 0
    box_lookup = defaultdict(list)
    while True:
        line = next(iterator)[1::4]
        if line.isnumeric():
            next(iterator)
            break

        for index, char in enumerate(line, 1):
            if char == " ":
                continue

            box_lookup[index].append(char)

    boxes = [box_lookup[i] for i in range(1, index + 1)]
    for stack in boxes:
        stack.reverse()

    movements = [
        Movement(int(line_info[0]), int(line_info[1]) - 1, int(line_info[2]) - 1)
        for line_info in (line.split()[1::2] for line in iterator)
    ]

    return boxes, movements


def work(boxes, movements, reverse=False):
    for movement in movements:
        append = boxes[movement.from_][-movement.count :]
        boxes[movement.to].extend(reversed(append) if reverse else append)
        del boxes[movement.from_][-movement.count :]


def run(data):
    base_boxes, movements = data

    for first in True, False:
        boxes = [stack.copy() for stack in base_boxes] if first else base_boxes
        work(boxes, movements, reverse=first)
        yield "".join(stack[-1] for stack in boxes)

File: csharp/2022/05.cs
namespace AoCSolutions.Year2022;
using Result = ValueTuple<List<string>, List<Movement>>;

public record struct Movement(byte Count, byte From, byte To);

public class Day05 : AoCRunner<Result>, IAoCDay
{
    public override (object?, object?) Run(Result data)
    {
        var (boxes, movements) = data;
        return (Work(boxes.ToList(), movements), Work(boxes, movements, reverse: false));
    }

    public static string Work(List<string> boxes, List<Movement> movements, bool reverse = true)
    {
        foreach (var movement in movements)
        {
            var append = boxes[movement.From][^movement.Count..];
            boxes[movement.To] += reverse ? String.Join("", append.Reverse()) : append;
            boxes[movement.From] = boxes[movement.From][..^movement.Count];
        }
        return String.Join("", boxes.Select(x => x[^1..]));
    }

    public override Result Process(IEnumerable<string> data)
    {
        var enumerator = data.GetEnumerator();
        List<string>? boxes = null;

        while (enumerator.MoveNext())
        {
            var boxLine = String.Join("", enumerator.Current.Skip(1).Step(4));
            boxes ??= Enumerable.Repeat("", boxLine.Length).ToList();

            if (boxLine.All(char.IsDigit))
            {
                enumerator.MoveNext();
                break;
            }

            foreach (var index in ..boxLine.Length)
            {
                var character = boxLine[index];
                if (character == ' ')
                    continue;

                boxes[index] = $"{character}{boxes[index]}";
            }
        }

        List<Movement> movements = new();

        while (enumerator.MoveNext())
        {
            var line = enumerator.Current.Split().Skip(1).Step(2).ToArray();
            movements.Add(
                new(
                    byte.Parse(line[0].ToString()),
                    (byte)(byte.Parse(line[1].ToString()) - 1),
                    (byte)(byte.Parse(line[2].ToString()) - 1)
                )
            );
        }

        return (boxes!, movements);
    }
}
